File: src/utils/validator.js
'use strict';

const { ErrorCodes, createError } = require('./error-codes');

// Forbidden key segments to prevent prototype pollution
const FORBIDDEN_SEGMENTS = new Set(['__proto__', 'constructor', 'prototype']);

// Maximum key length
const MAX_KEY_LENGTH = 1000;

// Maximum value size (10MB)
const MAX_VALUE_SIZE = 10 * 1024 * 1024;

// Control characters to reject
const CONTROL_CHARS = /[\x00-\x1F\x7F]/;

// Invisible characters to reject
const INVISIBLE_CHARS = /[\u200B-\u200D\uFEFF\u2028\u2029]/;

/**
 * Validate a key
 * @param {string} key - Key to validate
 * @returns {Array<string>} Array of key parts
 * @throws {Error} If key is invalid
 */
function validateKey(key) {
  if (typeof key !== 'string' || key.length === 0) {
    throw createError(ErrorCodes.INVALID_KEY, 'Key must be a non-empty string');
  }

  // Check key length
  if (key.length > MAX_KEY_LENGTH) {
    throw createError(
      ErrorCodes.KEY_TOO_LONG,
      `Key too long: ${key.length} chars (max: ${MAX_KEY_LENGTH})`
    );
  }

  // Normalize Unicode to NFC form
  const normalized = key.normalize('NFC');

  // Check for control characters
  if (CONTROL_CHARS.test(normalized)) {
    throw createError(ErrorCodes.KEY_CONTROL_CHARS, 'Key contains control characters');
  }

  // Check for invisible characters
  if (INVISIBLE_CHARS.test(normalized)) {
    throw createError(ErrorCodes.KEY_INVISIBLE_CHARS, 'Key contains invisible characters');
  }

  const parts = normalized.split('.');
  for (const part of parts) {
    if (part.length === 0) {
      throw createError(
        ErrorCodes.KEY_EMPTY_SEGMENT,
        `Key contains an empty segment: "${key}"`
      );
    }
    if (FORBIDDEN_SEGMENTS.has(part)) {
      throw createError(
        ErrorCodes.KEY_FORBIDDEN_SEGMENT,
        `Forbidden key segment "${part}" in key: "${key}"`
      );
    }
  }

  return parts;
}

/**
 * Validate a value
 * @param {*} value - Value to validate
 * @throws {Error} If value is invalid
 */
function validateValue(value) {
  // Check value size
  try {
    const size = JSON.stringify(value).length;
    if (size > MAX_VALUE_SIZE) {
      throw createError(
        ErrorCodes.VALUE_TOO_LARGE,
        `Value too large: ${size} bytes (max: ${MAX_VALUE_SIZE})`
      );
    }
  } catch (err) {
    if (err.message.includes('circular')) {
      throw createError(ErrorCodes.CIRCULAR_REFERENCE, 'Circular reference detected');
    }
    if (err.message.includes('BigInt')) {
      throw createError(ErrorCodes.UNSUPPORTED_TYPE, 'BigInt not supported');
    }
    throw err;
  }

  // Check for circular references
  const seen = new WeakSet();
  const stack = [value];

  while (stack.length > 0) {
    const current = stack.pop();

    if (typeof current === 'object' && current !== null) {
      if (seen.has(current)) {
        throw createError(ErrorCodes.CIRCULAR_REFERENCE, 'Circular reference detected');
      }
      seen.add(current);

      for (const key of Object.keys(current)) {
        stack.push(current[key]);
      }
    }
  }
}

/**
 * Validate key length
 * @param {string} key - Key to validate
 * @throws {Error} If key is too long
 */
function validateKeyLength(key) {
  if (key.length > MAX_KEY_LENGTH) {
    throw createError(
      ErrorCodes.KEY_TOO_LONG,
      `Key too long: ${key.length} chars (max: ${MAX_KEY_LENGTH})`
    );
  }
}

/**
 * Check if a key is sensitive (should be encrypted)
 * @param {string} key - Key to check
 * @returns {boolean} True if key is sensitive
 */
function isSensitiveKey(key) {
  // Updated regex to be more precise and avoid false positives
  const SENSITIVE_KEY_RE = /^(?:password|secret|token|apikey|api_key|private)(?:[._]|$)/i;
  return SENSITIVE_KEY_RE.test(key);
}

module.exports = {
  validateKey,
  validateValue,
  validateKeyLength,
  isSensitiveKey,
  FORBIDDEN_SEGMENTS,
  MAX_KEY_LENGTH,
  MAX_VALUE_SIZE,
};

File: crates/spectre-db-core/src/validator.rs
use crate::error::{codes, Result, SpectreError};
use unicode_normalization::UnicodeNormalization;

pub const MAX_KEY_LENGTH: usize = 1000;
pub const MAX_VALUE_SIZE: usize = 10 * 1024 * 1024;

const FORBIDDEN_SEGMENTS: [&str; 3] = ["__proto__", "constructor", "prototype"];

pub fn is_sensitive_key(key: &str) -> bool {
    let lower = key.to_ascii_lowercase();
    for prefix in ["password", "secret", "token", "apikey", "api_key", "private"] {
        if let Some(rest) = lower.strip_prefix(prefix) {
            if rest.is_empty() || rest.starts_with('.') || rest.starts_with('_') {
                return true;
            }
        }
    }
    false
}

pub fn validate_key(key: &str) -> Result<Vec<String>> {
    if key.is_empty() {
        return Err(SpectreError::invalid_key("Key must be a non-empty string"));
    }
    if key.chars().count() > MAX_KEY_LENGTH {
        return Err(SpectreError::new(
            codes::KEY_TOO_LONG,
            format!("Key too long: {} chars (max: {})", key.chars().count(), MAX_KEY_LENGTH),
        ));
    }

    let normalized: String = key.nfc().collect();

    for ch in normalized.chars() {
        if ch <= '\u{1F}' || ch == '\u{7F}' {
            return Err(SpectreError::new(codes::KEY_CONTROL_CHARS, "Key contains control characters"));
        }
        if matches!(ch, '\u{200B}'..='\u{200D}' | '\u{FEFF}' | '\u{2028}' | '\u{2029}') {
            return Err(SpectreError::new(codes::KEY_INVISIBLE_CHARS, "Key contains invisible characters"));
        }
    }

    let mut parts = Vec::new();
    for part in normalized.split('.') {
        if part.is_empty() {
            return Err(SpectreError::new(
                codes::KEY_EMPTY_SEGMENT,
                format!("Key contains an empty segment: \"{}\"", key),
            ));
        }
        if FORBIDDEN_SEGMENTS.contains(&part) {
            return Err(SpectreError::new(
                codes::KEY_FORBIDDEN_SEGMENT,
                format!("Forbidden key segment \"{}\" in key: \"{}\"", part, key),
            ));
        }
        parts.push(part.to_string());
    }
    Ok(parts)
}

pub fn validate_value_size(serialized_len: usize) -> Result<()> {
    if serialized_len > MAX_VALUE_SIZE {
        return Err(SpectreError::new(
            codes::VALUE_TOO_LARGE,
            format!("Value too large: {} bytes (max: {})", serialized_len, MAX_VALUE_SIZE),
        ));
    }
    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn key_validation() {
        assert!(validate_key("user.profile.name").is_ok());
        assert!(validate_key("").is_err());
        assert!(validate_key("a..b").is_err());
        assert!(validate_key("__proto__.x").is_err());
        assert!(validate_key("a\u{0000}b").is_err());
        assert!(validate_key(&"k".repeat(1001)).is_err());
    }

    #[test]
    fn sensitive_keys() {

        assert!(is_sensitive_key("password"));
        assert!(is_sensitive_key("password.holder"));
        assert!(is_sensitive_key("API_KEY.id"));
        assert!(is_sensitive_key("token"));
        assert!(!is_sensitive_key("user.password"));
        assert!(!is_sensitive_key("passwords"));
        assert!(!is_sensitive_key("username"));
        assert!(!is_sensitive_key("tokenBucket"));
    }

    #[test]
    fn nfc_normalization() {

        let parts = validate_key("caf\u{65}\u{301}.name").unwrap();
        assert_eq!(parts[0], "caf\u{E9}");
    }
}
